lib/active_merchant/billing/gateways/quickbooks.rb

```diff
@@ -170,30 +192,37 @@ def commit(uri, body = {}, method = :post)
         # The QuickBooks API returns HTTP 4xx on failed transactions, which causes a
         # ResponseError raise, so we have to inspect the response and discern between
         # a legitimate HTTP error and an actual gateway transactional error.
+        headers = {}
         response = begin
           case method
           when :post
-            ssl_post(endpoint, post_data(body), headers(:post, endpoint))
+            headers = headers(:post, endpoint)
+            ssl_post(endpoint, post_data(body), headers)
           when :get
-            ssl_request(:get, endpoint, nil, headers(:get, endpoint))
+            headers = headers(:get, endpoint)
+            ssl_request(:get, endpoint, nil, headers)
           else
             raise ArgumentError, "Invalid HTTP method: #{method}. Valid methods are :post and :get"
           end
         rescue ResponseError => e
           extract_response_body_or_raise(e)
         end
 
-        response_object(response)
+        response_object(response, headers)
       end
 
-      def response_object(raw_response)
+      def response_object(raw_response, headers = {})
         parsed_response = parse(raw_response)
 
+        if @options[:refresh_token]
+          parsed_response.merge!({ 'access_token' => @options[:access_token], 'refresh_token' => @options[:refresh_token] })
```

Comment:
After some end-to-end testing of the integration, made an update here to include refresh_token and access_token on the params object so that they can be updated on the gateway integration side of things as they change. Since access_tokens expire every hour, it's probably infeasible for a customer to update them on their end, so we need a way to actually save these updated values rather than continually refreshing with each unique transaction.

I am wondering if there are any concerns around saving these values on the response object though? Technically, without the client_id and client_secret, you couldn't use them to regenerate a token, and the access_token is only useful for a short span of time, but not sure if I'm overlooking any other possibilities. The other way I can think to get these values (for updating gateway settings on subsequent txns) is to grab them out of the txn's transcript before it's scrubbed.